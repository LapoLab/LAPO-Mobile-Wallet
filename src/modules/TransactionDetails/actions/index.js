// @flow
import { createActions } from 'redux-actions';

export const {
  saveContact,
  getContactFromAddress,
  setContactFromAddress,
  showContact,
  setLoadingContact,
} = createActions({
  SAVE_CONTACT: (address = '') => ({ address }),
  GET_CONTACT_FROM_ADDRESS: (address = '') => ({ address }),
  SET_CONTACT_FROM_ADDRESS: (name = '', id = '') => ({ name, id }),
  SHOW_CONTACT: (address = '') => ({ address }),
  SET_LOADING_CONTACT: (isLoading = true) => isLoading,
});

export { getContacts } from '../../ChooseContact/actions';
