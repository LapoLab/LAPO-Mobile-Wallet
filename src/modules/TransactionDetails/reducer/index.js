import { handleActions } from 'redux-actions';
import {
  setContactFromAddress,
  setLoadingContact,
} from '../actions';


export const initialChooseContactState = {
  nameContact: '',
  idContact: '',
  isLoadingContact: '',
};

// reducer
const reducer = handleActions(
  {
    [setContactFromAddress](state, { payload }) {
      return {
        ...state,
        nameContact: payload.name,
        idContact: payload.id,
      };
    },
    [setLoadingContact](state, { payload }) {
      return {
        ...state,
        isLoadingContact: payload,
      };
    },
  },
  initialChooseContactState,
);

export default reducer;
