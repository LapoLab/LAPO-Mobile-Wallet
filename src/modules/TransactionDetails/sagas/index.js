import saveContact from './saveContactSaga';
import showContact from './showContact';
import getContactFromAddress from './getContactFromAddress';
import showDetails from './showDetails';

export * from '../actions';

export default {
  saveContact,
  getContactFromAddress,
  showContact,
  showDetails,
};
