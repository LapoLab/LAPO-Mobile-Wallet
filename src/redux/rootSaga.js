import { all } from 'redux-saga/effects';
import commonsSagas from '../commons/sagas';
import recoveryWordsSagas from '../modules/RecoveryWords/sagas';
import sendSagas from '../modules/Send/sagas';
import paperWordsConfirm from '../modules/PaperWordsConfirm/sagas';
import dashboard from '../modules/Dashboard/sagas';
import settings from '../modules/Settings/sagas';
import paperWords from '../modules/PaperWords/sagas';
import transactionsDetails from '../modules/TransactionDetails/sagas';
import chooseContact from '../modules/ChooseContact/sagas';
import splashScreen from '../modules/SplashScreen/sagas';
import switchContacts from '../modules/SwitchContacts/sagas';
import receive from '../modules/Receive/sagas';

// eslint-disable-next-line
const getListeners = (...args) => args.reduce((acc, nextArg) => [...acc, ...Object.values(nextArg).map(func => func())], []);

export default function* rootSaga() {
  yield all(
    getListeners(
      commonsSagas,
      recoveryWordsSagas,
      sendSagas,
      paperWordsConfirm,
      dashboard,
      settings,
      paperWords,
      transactionsDetails,
      chooseContact,
      splashScreen,
      switchContacts,
      receive,
    ),
  );
}
