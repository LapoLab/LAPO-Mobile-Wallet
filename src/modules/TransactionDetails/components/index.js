/**
 * @flow
 */

import React, { Component } from 'react';
import {
  Text,
  View,
  TouchableHighlight,
  Image,
  TouchableOpacity,
  ActivityIndicator,
} from 'react-native';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import Share from 'react-native-share';
import moment from 'moment';
import calcAmount from '../../../utils/calcAmount';
import * as actions from '../actions';
import Button from '../../../commons/components/Button';
import * as actionsCommon from '../../../commons/actions';
import styles from './styles';
import { close, copyIcon } from '../../../assets';

type Props = {
  saveContact: Function,
  isLoaded: boolean,
  isReady: boolean,
  onClose: Function,
  note: string,
  nameContact: string,
  isLoadingContact: boolean,
  transaction: { address: string, amount: number, time: number, mode: string, status: string },
  getContactFromAddress: Function,
  showContact: Function,
  goToOtherWindow: Function,
};
class TransactionDetails extends Component<Props> {
  componentDidMount() {
    const { getContactFromAddress } = this.props;
    getContactFromAddress();
  }

  svg = null;

  handleClose = () => {
    const { onClose } = this.props;
    onClose();
  };

  shareOnlyAddress = () => {
    const {
      transaction: { address },
      goToOtherWindow,
    } = this.props;

    goToOtherWindow(true);

    const shareImageBase64 = {
      title: 'Lapo address',
      message: address || '',
    };

    // eslint-disable-next-line
    Share.open(shareImageBase64).catch(error => console.log(error));
  };

  handleAddContacts = () => {
    const { saveContact, transaction } = this.props;
    const { address } = transaction || { address: '' };
    saveContact(address);
  };

  handleOpenContact = () => {
    const { showContact, transaction } = this.props;
    const { address } = transaction || { address: '' };

    showContact(address);
  };

  render() {
    const {
      transaction, note, isLoaded, isReady, nameContact, isLoadingContact,
    } = this.props;
    const {
      address, amount, time, mode, status,
    } = transaction || {
      address: '',
      amount: '',
      time: '',
      mode: '',
      status: '',
    };

    return (
      <TouchableHighlight
        style={styles.GlobalContainer}
        onPress={this.handleClose}
        underlayColor="transparent"
      >
        <TouchableHighlight style={styles.container} underlayColor="#EFF4FC">
          <View style={{ flex: 1, width: '100%' }}>
            <View style={styles.header}>
              <TouchableOpacity onPress={this.handleClose} style={styles.buttonClose}>
                <Image style={styles.headerIcon} source={close} />
              </TouchableOpacity>
              <Text style={styles.headerTitle}>Transaction details</Text>
              <TouchableOpacity style={styles.buttonClose} />
            </View>

            <View style={styles.row}>
              <Text style={styles.rowTitle}>{mode === 'out' ? 'From' : 'To'}</Text>
              <Text style={styles.valueAddress}>{address}</Text>
              <TouchableOpacity onPress={this.shareOnlyAddress}>
                <Image style={styles.copyIcon} source={copyIcon} />
              </TouchableOpacity>
            </View>

            <View style={styles.row}>
              <Text style={styles.rowTitle}>Amount</Text>
              <Text
                style={[styles.valueParam, mode === 'out' ? { color: 'green' } : { color: 'red' }]}
              >
                {mode === 'out' ? '+  ' : '-  '}
                {calcAmount.calcAmountLux(amount)}
                {'  LAX'}
              </Text>
            </View>

            <View style={styles.row}>
              <Text style={styles.rowTitle}>Status</Text>
              <Text style={styles.valueParam}>{isLoaded && isReady ? status : '-'}</Text>
            </View>

            <View style={styles.row}>
              <Text style={styles.rowTitle}>Date</Text>
              <Text style={styles.valueParam}>
                {time > 154262948 ? moment.unix(time).format('DD MMM YYYY - HH:mm') : '-'}
              </Text>
            </View>

            <View style={styles.row}>
              <Text style={styles.rowTitle}>Note</Text>
              <Text style={styles.valueParam}>{note}</Text>
            </View>

            <View style={styles.row}>
              <Text style={styles.rowTitle}>Contact</Text>
              <Text style={styles.valueParam}>{nameContact}</Text>
            </View>

            {!isLoadingContact && (
              <Button
                onPress={nameContact === '' ? this.handleAddContacts : this.handleOpenContact}
                title={nameContact === '' ? 'ADD TO CONTACTS' : 'VIEW CONTACT'}
                status={nameContact === '' ? 'active' : 'gray'}
              />
            )}
            {isLoadingContact ? (
              <View style={styles.containerSpiner}>
                <ActivityIndicator size="large" color="#565656" style={styles.spinner} />
              </View>
            ) : null}

          </View>
        </TouchableHighlight>
      </TouchableHighlight>
    );
  }
}

const mapStateToProps = state => ({
  isShowSpinner: state.modals.isShowSpinner,
  isLoaded: state.dashboard.maxBlock - state.dashboard.countBlock < 10,
  isReady: state.dashboard.isReady,
  txDetails: state.modals.txDetails || '',
  transaction:
    state.wallet.transactions
    && state.wallet.transactions.find(item => item.id === state.modals.txDetails),
  note: state.wallet.notes[state.modals.txDetails] || '',
  nameContact: state.transactionsDetails.nameContact,
  isLoadingContact: state.transactionsDetails.isLoadingContact,
});

const mapDispatchToProps = dispatch => bindActionCreators(
  {
    saveContact: actions.saveContact,
    showContact: actions.showContact,
    getContactFromAddress: actions.getContactFromAddress,
    goToOtherWindow: actionsCommon.goToOtherWindow,
    // setModalSwitchContacts: actionsCommon.setModalSwitchContacts,
  },
  dispatch,
);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TransactionDetails);
