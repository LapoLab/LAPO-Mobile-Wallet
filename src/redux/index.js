import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from 'redux-saga';
import storage from 'redux-persist/lib/storage';
import { persistStore, persistReducer } from 'redux-persist';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';
import rootReducer from './rootReducer';
import rootSagas from './rootSaga';

const persistConfig = {
  key: 'root',
  storage,
  stateReconciler: autoMergeLevel2,
  blacklist: [
    'modals',
    'dashboard',
    'paperWords',
    'chooseContact',
    'send',
    'transactionsDetails',
    'alert',
    'pinsRecovery',
  ],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const sagaMiddleware = createSagaMiddleware();
// eslint-disable-next-line no-undef
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export const store = createStore(
  persistedReducer,
  composeEnhancers(applyMiddleware(sagaMiddleware)),
);
sagaMiddleware.run(rootSagas);

export const persistor = persistStore(store);
