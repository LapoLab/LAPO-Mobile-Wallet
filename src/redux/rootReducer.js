import { combineReducers } from 'redux';

import wallet from '../commons/reducers/wallet';
import pins from '../commons/reducers/pins';
import pinsRecovery from '../commons/reducers/pinsRecovery';
import modals from '../commons/reducers/modals';
import settings from '../commons/reducers/settings';
import notify from '../modules/Notify/reducers';
import paperConfirm from '../modules/PaperWordsConfirm/reducer';
import dashboard from '../modules/Dashboard/reducers';
import paperWords from '../modules/PaperWords/reducers';
import chooseContact from '../modules/ChooseContact/reducers';
import send from '../modules/Send/reducers';
import transactionsDetails from '../modules/TransactionDetails/reducer';
import alert from '../modules/Alert/reducers';

export default combineReducers({
  wallet,
  pins,
  pinsRecovery,
  notify,
  modals,
  settings,
  paperConfirm,
  dashboard,
  paperWords,
  chooseContact,
  send,
  transactionsDetails,
  alert,
});
