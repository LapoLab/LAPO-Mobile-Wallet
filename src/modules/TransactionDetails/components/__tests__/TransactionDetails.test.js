// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from '../../../../redux';
import TransactionDetails from '../index';

describe('TransactionDetails page', () => {
  test('renders snapshots', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <TransactionDetails />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
