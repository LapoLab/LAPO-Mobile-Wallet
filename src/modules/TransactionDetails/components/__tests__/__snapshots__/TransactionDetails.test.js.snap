// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionDetails page renders snapshots 1`] = `
<View
  accessible={true}
  isTVSelectable={true}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "flex-end",
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "flex-start",
        "backgroundColor": "#EFF4FC",
        "borderTopLeftRadius": 12,
        "borderTopRightRadius": 12,
        "borderWidth": 0,
        "height": 510,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#D8D8D8",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "space-between",
            "paddingRight": 15,
            "paddingVertical": 15,
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "padding": 15,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/close.png",
              }
            }
            style={
              Object {
                "height": 15,
                "resizeMode": "contain",
                "width": 15,
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "color": "#131313",
              "flex": 1,
              "fontFamily": "Lato-Bold",
              "fontSize": 18,
              "textAlign": "center",
            }
          }
        >
          Transaction details
        </Text>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "padding": 15,
            }
          }
        />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#D8D8D8",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "space-between",
            "paddingHorizontal": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#565656",
              "fontFamily": "Lato-Bold",
              "fontSize": 16,
              "lineHeight": 24,
              "width": 70,
            }
          }
        >
          To
        </Text>
        <Text
          style={
            Object {
              "color": "black",
              "flex": 1,
              "fontFamily": "Lato-Regular",
              "fontSize": 8,
              "marginBottom": 6,
              "marginLeft": 10,
              "marginTop": 8,
              "textAlign": "left",
            }
          }
        >
          
        </Text>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/copyIcon.png",
              }
            }
            style={
              Object {
                "resizeMode": "contain",
                "width": 20,
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#D8D8D8",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "space-between",
            "paddingHorizontal": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#565656",
              "fontFamily": "Lato-Bold",
              "fontSize": 16,
              "lineHeight": 24,
              "width": 70,
            }
          }
        >
          Amount
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "black",
                "flex": 1,
                "fontFamily": "Lato-Regular",
                "fontSize": 16,
                "marginBottom": 8,
                "marginLeft": 10,
                "marginTop": 8,
                "textAlign": "left",
              },
              Object {
                "color": "red",
              },
            ]
          }
        >
          -  
          0
            LAX
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#D8D8D8",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "space-between",
            "paddingHorizontal": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#565656",
              "fontFamily": "Lato-Bold",
              "fontSize": 16,
              "lineHeight": 24,
              "width": 70,
            }
          }
        >
          Status
        </Text>
        <Text
          style={
            Object {
              "color": "black",
              "flex": 1,
              "fontFamily": "Lato-Regular",
              "fontSize": 16,
              "marginBottom": 8,
              "marginLeft": 10,
              "marginTop": 8,
              "textAlign": "left",
            }
          }
        >
          -
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#D8D8D8",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "space-between",
            "paddingHorizontal": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#565656",
              "fontFamily": "Lato-Bold",
              "fontSize": 16,
              "lineHeight": 24,
              "width": 70,
            }
          }
        >
          Date
        </Text>
        <Text
          style={
            Object {
              "color": "black",
              "flex": 1,
              "fontFamily": "Lato-Regular",
              "fontSize": 16,
              "marginBottom": 8,
              "marginLeft": 10,
              "marginTop": 8,
              "textAlign": "left",
            }
          }
        >
          -
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#D8D8D8",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "space-between",
            "paddingHorizontal": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#565656",
              "fontFamily": "Lato-Bold",
              "fontSize": 16,
              "lineHeight": 24,
              "width": 70,
            }
          }
        >
          Note
        </Text>
        <Text
          style={
            Object {
              "color": "black",
              "flex": 1,
              "fontFamily": "Lato-Regular",
              "fontSize": 16,
              "marginBottom": 8,
              "marginLeft": 10,
              "marginTop": 8,
              "textAlign": "left",
            }
          }
        >
          
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomColor": "#D8D8D8",
            "borderBottomWidth": 1,
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "space-between",
            "paddingHorizontal": 15,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#565656",
              "fontFamily": "Lato-Bold",
              "fontSize": 16,
              "lineHeight": 24,
              "width": 70,
            }
          }
        >
          Contact
        </Text>
        <Text
          style={
            Object {
              "color": "black",
              "flex": 1,
              "fontFamily": "Lato-Regular",
              "fontSize": 16,
              "marginBottom": 8,
              "marginLeft": 10,
              "marginTop": 8,
              "textAlign": "left",
            }
          }
        >
          
        </Text>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
          }
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#4183BD",
              "borderColor": "#BBCDEB",
              "borderRadius": 74,
              "borderWidth": 8,
              "height": 64,
              "justifyContent": "center",
              "marginHorizontal": 25,
              "opacity": 1,
              "width": "80%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontFamily": "Lato-Regular",
                "fontSize": 18,
              }
            }
          >
            ADD TO CONTACTS
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
